### PR TITLE
Use mkl gemm batch in CPU backend when appropriate

### LINKED ISSUE #2146: Take advantage of MKL batch gemm
Hello everybody
I ve seen this article on Intel website :  
https://software.intel.com/en-us/articles/introducing-batch-gemm-operations
Introducing Batch GEMM Operations
...
Intel MKL 11.3 Beta (part of Intel® Parallel Studio XE 2016 Beta) includes a new flavor of GEMM feature called "Batch GEMM". This allows users to achieve the same objective described above with minimal programming effort. Users can specify multiple independent GEMM operations, which can be of different matrix sizes and different parameters, through a single call to the "Batch GEMM" API. At runtime, Intel MKL will intelligently execute all of the matrix multiplications so as to optimize overall performance. Here is an example that shows how "Batch GEMM" works:
....

I have searched in the AF sources and dont see any "cblas_dgemm_batch".
I remember that CLBlast is able of batching matmul. 
But is afcpu (3.6.0) capable of batching matmul ?
Kind
W.

### PR DESCRIPTION
**Pull request by 9prady9**
Resolves #2146

### CODE REVIEW

**Comment by 9prady9** [comment]
@WilliamTambellini Yes, I am working on benchmarks. Once, I have them in presentable format, I will share them here.

**Comment by 9prady9** [comment]
Sorry about the delay but we wanted to make sure the changes are indeed giving the performance improvements we expected it to. Turns out, if we Intel TBB as threading solution, MKL batch call wasn't giving any performance improvements for some reason. Once we switched the threading to be `GNU OpenMP`, we started seeing good performance improvements.

Given below is the link to the interactive chart that shows the numbers we have for batched operations.
https://docs.google.com/spreadsheets/d/e/2PACX-1vRfKVz-VV9qGae7fLC3wYHtvNROwgfZ9yI7mCDTvxXJ2wqV8ibtrp1BVxykIz9nTVlCg5ouRvfd1hFN/pubchart?oid=1109307214&format=interactive

It is by no means a comprehensive set of benchmarks but serves as an confirmation of the potential performance improvements and issue with TBB.

**Comment by 9prady9** [comment]
looks like batched matmul failed, I will look into it.

**Comment by WilliamTambellini** [comment]
Hum very interesting. Congrats.
@umar456  I am aware about another drawback with TBB backend : it seems impossible to finely select the number of cpu cores to be used by afcpu. With OpenMP, just setting OMP_NUM_THREADS was at least working with AF 3.5.  I know afcpu 3.6.1 has been built with TBB backend but would it be possible for afcpu 3.6.2 to switch back to OpenMP (gnu or intel, your choice) ?

**Comment by 9prady9** [comment]
We will be defaulting to `GNU OpenMP` on non-Windows platforms and `Intel OpenMP` on windows.

**Comment by 9prady9** [comment]
The current changes to FindMKL seems to work on both Linux and Windows. However, OSX still can't find OpenMP, looking into it.

**Comment by 9prady9** [comment]
`linux-cpu-e5`(given below are the CPU details) seems be giving a slightly higher error value of `9.15527e-05` compared to the expected `1e-05`. I have changed the expected value to be `1e-04`, @umar456 let me know you don't like that.

```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 61
model name	: Intel Core Processor (Broadwell, IBRS)
siblings	: 4
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl eagerfpu pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch arat xsaveopt invpcid_single tpr_shadow vnmi flexpriority ept vpid fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap
```

**Comment by 9prady9** [comment]
@umar456 I am not sure how to proceed, the blas_cpu test is failing via ci on `linux-cpu-e5` with Intel OpenMP, however it doesn't fail when I run it manually from shell. This is the only blocker on this PR at the moment. All other jobs seem to be running fine with Intel OpenMP as thread layer.

File: CMakeModules/FindMKL.cmake [hunk 1]
```diff
@@ -219,10 +219,20 @@ elseif(MKL_THREAD_LAYER STREQUAL "GNU OpenMP")
   find_package(OpenMP REQUIRED)
   find_mkl_library(NAME ThreadLayer LIBRARY_NAME mkl_gnu_thread)
   add_library(MKL::ThreadingLibrary SHARED IMPORTED)
+  set(omp_lib_var "")
+  foreach(lib_name ${OpenMP_CXX_LIB_NAMES})
```
[the remark this answers is not in the document]
Comment (9prady9) [reply CMakeModules/FindMKL.cmake]: When I originally fixed this, the FindMKL script wasn't fixed for this case - fixed later in an individual PR.

Also, I had Windows and OSX on my mind where OMP library names are different. To address that I used this format. I was treating `GNU OpenMP` as openmp that comes along with native compiler.

File: CMakeLists.txt [hunk 2]
```diff
@@ -22,6 +22,9 @@ include(CheckCXXCompilerFlag)
 
 arrayfire_set_cmake_default_variables()
 
+#Set Intel OpenMP as default MKL thread layer
+set(MKL_THREAD_LAYER "Intel OpenMP" CACHE STRING "The thread layer to choose for MKL")
```
Comment (umar456) [inline CMakeLists.txt]: This should go in the FindMKL.cmake file.

**Pull request merged**